srml/im-online/src/lib.rs

```diff
@@ -233,24 +233,20 @@ decl_module! {
 		fn heartbeat(
 			origin,
 			heartbeat: Heartbeat<T::BlockNumber>,
-			signature: <T::AuthorityId as RuntimeAppPublic>::Signature
+			// since signature verification is done in `validate_unsigned`
+			// we can skip doing it here again.
+			_signature: <T::AuthorityId as RuntimeAppPublic>::Signature
 		) {
 			ensure_none(origin)?;
 
 			let current_session = <session::Module<T>>::current_index();
-			ensure!(current_session == heartbeat.session_index, "Outdated heartbeat received.");
 			let exists = <ReceivedHeartbeats>::exists(
 				&current_session,
 				&heartbeat.authority_index
 			);
 			let keys = Keys::<T>::get();
 			let public = keys.get(heartbeat.authority_index as usize);
 			if let (true, Some(public)) = (!exists, public) {
-				let signature_valid = heartbeat.using_encoded(|encoded_heartbeat| {
```

Comment:
why is this gone?

Reply:
It's done in `validate_unsigned` as well before dispatching the transaction. @tomusdrw are `pre_dispatch` checks run as part of block execution? Otherwise someone could include an invalid transaction into a block if the signature is only checked there.

Reply:
Yes, that's the main difference between `validate_transaction`(run only by the pool, to determine if transaction should enter it or not) and `pre_dispatch` run during block processing to make sure the transaction is valid. A transaction with failing `pre_dispatch` should cause a block import to panic. We are only able to recover from it during block authorship (we just skip transaction, mark it as invalid in the pool and try others).

The reason why we have pre-dispatch failing transactions in the pool is asynchrony between `validate_transaction` and the actual block creation. Some transactions might become invalid because some other transactions were imported first, and such relationship is not possible to express via `provides`.

Reply:
Thanks for the explanation! :)